In [5]:
import numpy as np
import pandas as pd

In [ ]:
dataset = pd.read_xlsx("1_18150893050.xlsx")
dataset.to_csv("1_18150893050.csv")

read the file and covert it to csv

In [6]:
dataset = pd.read_csv("1_18150893050.csv",index_col="Unnamed: 0" )
data_array = dataset.to_numpy()


read the csv file and creat a numpy array

In [7]:
data_df = pd.DataFrame(data_array , columns=["InvoiceNo","StockCode","Description","Quantity","InvoiceDate","UnitPrice","CustomerID","Country"])

creat a data frame

In [8]:
print(data_df)

       InvoiceNo StockCode                          Description Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER        6   
1         536365     71053                  WHITE METAL LANTERN        6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER        8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE        6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.        6   
...          ...       ...                                  ...      ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS       12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL         6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL         4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE        4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT         3   

                InvoiceDate UnitPrice CustomerID         Country  
0       2010-12-01 0

In [9]:
print(data_df.info())
print(data_df.isna().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype 
---  ------       --------------   ----- 
 0   InvoiceNo    541909 non-null  object
 1   StockCode    541909 non-null  object
 2   Description  540455 non-null  object
 3   Quantity     541909 non-null  object
 4   InvoiceDate  541909 non-null  object
 5   UnitPrice    541909 non-null  object
 6   CustomerID   406829 non-null  object
 7   Country      541909 non-null  object
dtypes: object(8)
memory usage: 33.1+ MB
None
InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64


in this file, there is 541909 samples.all are object. in two columns: Description,CustomerID there is 1454 and 135080 missing data

In [10]:
data_df["Quantity"] = pd.to_numeric(data_df["Quantity"], errors="coerce")
data_df["CustomerID"] = pd.to_numeric(data_df["CustomerID"], errors="coerce")
data_df["UnitPrice"] = pd.to_numeric(data_df["UnitPrice"], errors="coerce")
data_df['InvoiceDate'] = pd.to_datetime(data_df['InvoiceDate'])

change Quantity,CustomerID,UnitPrice type to float.

In [11]:
data_df["CustomerID"] = data_df["CustomerID"].fillna("unknown")

fill null datas in CustomerID to unknown.

In [12]:
data_df = data_df.dropna(subset=["Description"])

deleting samples which are null in Description column.

In [13]:
print(data_df.info())
print(data_df.isna().sum())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 540455 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    540455 non-null  object        
 1   StockCode    540455 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     540455 non-null  int64         
 4   InvoiceDate  540455 non-null  datetime64[ns]
 5   UnitPrice    540455 non-null  float64       
 6   CustomerID   540455 non-null  object        
 7   Country      540455 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 37.1+ MB
None
InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64


In [14]:
data_df["Quantity"].where(data_df["Quantity"]>500).dropna()

730         600.0
3917        600.0
4850       1824.0
4945       2880.0
4946       1400.0
           ...   
537883      960.0
540061      698.0
540070     1500.0
540071     1200.0
540421    80995.0
Name: Quantity, Length: 423, dtype: float64

all transactions with negative quantities

In [15]:
data_df["UnitPrice"].where(data_df["UnitPrice"] == 0).dropna()

6391      0.0
6392      0.0
7313      0.0
9302      0.0
13217     0.0
         ... 
535336    0.0
536908    0.0
538504    0.0
538505    0.0
538919    0.0
Name: UnitPrice, Length: 1061, dtype: float64

transactions where the unit price is zero.

In [16]:
data_df["Description"].where(data_df["Description"].str.contains("gift", case=False, na=False)).dropna()

30                         STARS GIFT TAPE 
712               PACK OF 6 BIRDY GIFT TAGS
836            PACK OF 6 SWEETIE GIFT BOXES
839               PACK OF 6 BIRDY GIFT TAGS
950       SMALL STRIPES CHOCOLATE GIFT BAG 
                        ...                
541502        50'S CHRISTMAS GIFT BAG LARGE
541676       6 GIFT TAGS VINTAGE CHRISTMAS 
541794            PACK OF 6 BIRDY GIFT TAGS
541813       6 GIFT TAGS VINTAGE CHRISTMAS 
541859          TEA TIME TEAPOT IN GIFT BOX
Name: Description, Length: 5293, dtype: object

all products whose description includes the word "gift"

In [17]:
data_df["Country"].where(data_df["Country"] == "Germany").dropna()

1109      Germany
1110      Germany
1111      Germany
1112      Germany
1113      Germany
           ...   
541801    Germany
541802    Germany
541803    Germany
541804    Germany
541805    Germany
Name: Country, Length: 9495, dtype: object

all orders placed from Germany

In [18]:
data_df.loc[1000:1999, ["InvoiceDate"]].where(data_df["InvoiceDate"].dt.month == 12).dropna()

InvoiceDate
1000 2010-12-01 12:43:00
1001 2010-12-01 12:43:00
1002 2010-12-01 12:43:00
1003 2010-12-01 12:43:00
1004 2010-12-01 12:43:00
...                  ...
1995 2010-12-01 14:34:00
1996 2010-12-01 14:34:00
1997 2010-12-01 14:34:00
1998 2010-12-01 14:34:00
1999 2010-12-01 14:34:00

[995 rows x 1 columns]

all orders in December (month 12) from rows 1000 to 1999.

In [19]:
lambda Quantity,UnitPrice : UnitPrice*Quantity

<function __main__.<lambda>(Quantity, UnitPrice)>

In [21]:
data_df["TotalCust"] = data_df.apply(lambda row: row["Quantity"] * row["UnitPrice"], axis=1)
print(data_df[["UnitPrice","Quantity","TotalCust"]])

        UnitPrice  Quantity  TotalCust
0            2.55         6      15.30
1            3.39         6      20.34
2            2.75         8      22.00
3            3.39         6      20.34
4            3.39         6      20.34
...           ...       ...        ...
541904       0.85        12      10.20
541905       2.10         6      12.60
541906       4.15         4      16.60
541907       4.15         4      16.60
541908       4.95         3      14.85

[540455 rows x 3 columns]
